contracts/ConstAddressDeployer.sol

```diff
@@ -3,34 +3,61 @@
 pragma solidity ^0.8.0;
 
 contract ConstAddressDeployer {
+    error EmptyBytecode();
+    error EmptyInit();
+    error FailedDeploy();
     error FailedInit();
 
-    event Deployed(bytes32 bytecodeHash, bytes32 salt, address deployedAddress);
+    event Deployed(bytes32 indexed bytecodeHash, bytes32 indexed salt, address indexed deployedAddress);
 
+    /**
+     * @dev Deploys a contract using `CREATE2`. The address where the contract
+     * will be deployed can be known in advance via {deployedAddress}.
+     *
+     * The bytecode for a contract can be obtained from Solidity with
+     * `type(contractName).creationCode`.
+     *
+     * Requirements:
+     *
+     * - `bytecode` must not be empty.
+     * - `salt` must have not been used for `bytecode` already by the same `msg.sender`.
+     */
     function deploy(bytes memory bytecode, bytes32 salt) external returns (address deployedAddress_) {
-        bytes32 newSalt = keccak256(abi.encode(msg.sender, salt));
-        // solhint-disable-next-line no-inline-assembly
-        assembly {
-            deployedAddress_ := create2(0, add(bytecode, 32), mload(bytecode), newSalt)
-        }
-        emit Deployed(keccak256(bytecode), salt, deployedAddress_);
+        deployedAddress_ = _deploy(bytecode, msg.sender, salt);
     }
 
+    /**
+     * @dev Deploys a contract using `CREATE2` and initialize it. The address where the contract
+     * will be deployed can be known in advance via {deployedAddress}.
+     *
+     * The bytecode for a contract can be obtained from Solidity with
+     * `type(contractName).creationCode`.
+     *
+     * Requirements:
+     *
+     * - `bytecode` must not be empty.
+     * - `salt` must have not been used for `bytecode` already by the same `msg.sender`.
+     * - `init` must not be empty. It is used to initialize the deployed contract
+     *    as an option to not have the constructor args affect the address derived by `CREATE2`.
+     */
     function deployAndInit(
         bytes memory bytecode,
         bytes32 salt,
         bytes calldata init
     ) external returns (address deployedAddress_) {
-        bytes32 newSalt = keccak256(abi.encode(msg.sender, salt));
-        // solhint-disable-next-line no-inline-assembly
-        assembly {
-            deployedAddress_ := create2(0, add(bytecode, 32), mload(bytecode), newSalt)
-        }
-        (bool success, ) = deployedAddress_.call(init);
-        if (!success) revert FailedInit();
-        emit Deployed(keccak256(bytecode), salt, deployedAddress_);
+        if (init.length == uint256(0)) revert EmptyInit();
+
+        deployedAddress_ = _deploy(bytecode, msg.sender, salt);
```

Comment:
```suggestion
        deployedAddress_ = _deploy(bytecode, msg.sender, keccak256(abi.encode(sender, salt)));
```

Reply:
we can remove the `sender` argument from _deploy and pass the hash as you suggest here but passing both the sender and the hash of the sender and the salt is redundant.

Reply:
I didn't realize that sender is only used to calculate the salt down the line. Sure.